### DISCUSSION TITLE
Ocpp Charger Simulator

### DISCUSSION THREAD

**Discussion by drc38** [lbbrhzn/ocpp#1339] [Ideas]
A useful tool for anyone wanting to test the integration without a real charger, or during development testing of the integration:
https://github.com/matth-x/MicroOcppSimulator

**Comment by lbbrhzn** [comment]
Cool. Would be great if we could add some support for this in our repo.

**Reply by drc38** [reply]
I thought so initially, not sure how it could be achieved though. To get the websocket connection to work on the same host I needed to create a user-defined bridge https://docs.docker.com/engine/network/drivers/bridge/

**Reply by purcell-lab** [reply]
> To get the websocket connection to work on the same host

Can you use two hosts on the same network?

One running the OCPP Charge Point Simulator and the other running this integration to talk to the Charge point simulator?

**Reply by BJReplay** [reply]
> Can you use two hosts on the same network?

Trying now.  Will report back.

**Reply by BJReplay** [reply]
> Can you use two hosts on the same network?
> 
> One running the OCPP Charge Point Simulator and the other running this integration to talk to the Charge point simulator?

Same network? Sort of.  Same host - not yet, and not really so far on the same network.

Issues so far:

I haven't delved into the code to work out how to get it working without a secure connection.

Without a secure connection, 

```
I got [MO] warning (MicroOcppMongooseClient.cpp:700): Insecure connection (WS)
[MO] info (MicroOcppMongooseClient.cpp:722): connection ws://ha.hostname:9000/charger -- closed
```
Logged every minute.  Trying to work out whether it needed a secure connection, I burned a lot of time trying to get secure connections up and running, and I've got OCPP listening, but HA is currently off the air.  I blame my abysmal Caddy skills.

I'm running caddy for a local connection for HA just so I don't have to deal with browsers warning about insecure websites, and that has worked well, until now, but trying to get WSS working has been **interesting**.  It seems to be working, but I've hit the latest brick wall which will leave me giving up for the moment, this evening:

On the HA side (HA running in docker), I'm seeing this logged each minute:

```
homeassistant   | 2024-11-26 21:39:31.044 WARNING (MainThread) [custom_components.ocpp] Protocols mismatched | expected Subprotocols: ['ocpp1.6', 'ocpp2.0.1'], but client supports   | Closing connection
homeassistant   | 2024-11-26 21:40:32.627 WARNING (MainThread) [custom_components.ocpp] Protocols mismatched | expected Subprotocols: ['ocpp1.6', 'ocpp2.0.1'], but client supports   | Closing connection
homeassistant   | 2024-11-26 21:42:31.079 WARNING (MainThread) [custom_components.ocpp] Protocols mismatched | expected Subprotocols: ['ocpp1.6', 'ocpp2.0.1'], but client supports   | Closing connection
```

On the MicroOcppSimulator, I'm seeing this logged each minute:

```
[MO] info (MicroOcppMongooseClient.cpp:722): connection wss://ha.hostname:9000/charger -- closed
[MO] info (MicroOcppMongooseClient.cpp:722): connection wss://ha.hostname:9000/charger -- closed
[MO] info (MicroOcppMongooseClient.cpp:722): connection wss://ha.hostname:9000/charger -- closed
```

So, they're definitely talking, but not particularly well.  Same network, different docker hosts, but it's a start.

Now to revert my config to get HA back up and running.

**Comment by purcell-lab** [comment]
I have installed MicrOCPP Simulator on one host and Home Assistant on another host with the OCPP integration and they are talking to each other nicely, even over insecure connection.

![image](https://github.com/user-attachments/assets/d973726c-043c-415d-a38d-00eb18558608)

![image](https://github.com/user-attachments/assets/44d667cd-eb6a-4984-b459-e6a6c9ece284)

The trick was updating the WS address to reflect the OCPP integration ws://192.168.86.54:9000

![image](https://github.com/user-attachments/assets/3d54e41d-9bfe-4401-a0fe-3da810b4043b)

![image](https://github.com/user-attachments/assets/a5a17c3c-1148-4876-82ca-1c71a76f9267)

```logs
[MO] Recv: [3,"730543b5-365e-bca4-c65e-bca5a3561af6",{}]
[MO] info (Connector.cpp:308): Session mngt: timeout
[MO] info (StatusNotification.cpp:52): New status: Available (connectorId 1)
[MO] Send: [2,"a35574ad-bdd2-b513-5e31-b3c62b0d7b3e","StatusNotification",{"connectorId":1,"errorCode":"NoError","status":"Available","timestamp":"2024-11-27T00:34:38.156Z"}]
[MO] Recv: [3,"a35574ad-bdd2-b513-5e31-b3c62b0d7b3e",{}]
[MO] info (StatusNotification.cpp:52): New status: Preparing (connectorId 1)
[MO] Send: [2,"730543b6-bfd0-2d5c-49d6-ce62efeaa7f2","StatusNotification",{"connectorId":1,"errorCode":"NoError","status":"Preparing","timestamp":"2024-11-27T00:34:44.270Z"}]
[MO] Recv: [3,"730543b6-bfd0-2d5c-49d6-ce62efeaa7f2",{}]
[MO] Recv: [2,"db9e5b83-5edc-4b89-b58a-75d34c90ab6f","RemoteStartTransaction",{"idTag":"MCBCRBIS4XMIUZKY01FB","connectorId":1}]
[MO] Send: [3,"db9e5b83-5edc-4b89-b58a-75d34c90ab6f",{"status":"Accepted"}]
[MO] info (Connector.cpp:341): Session mngt: trigger StartTransaction
[MO] info (SmartChargingService.cpp:168): New limit for connector 1, scheduled at = 2024-11-27T00:35:51.814Z, nextChange = 2037-01-01T00:00:00.000Z, limit = {-1.0, -1.0, -1}
[MO] Send: [2,"c879f091-782c-6395-3450-9d6c755326ca","StartTransaction",{"connectorId":1,"idTag":"MCBCRBIS4XMIUZKY01FB","meterStart":0,"timestamp":"2024-11-27T00:35:51.814Z"}]
[MO] info (StatusNotification.cpp:52): New status: Charging (connectorId 1)
[MO] Recv: [3,"c879f091-782c-6395-3450-9d6c755326ca",{"transactionId":1732667752,"idTagInfo":{"status":"Accepted"}}]
[MO] info (StartTransaction.cpp:65): Request has been accepted
[MO] Send: [2,"8712bef2-f7a9-8128-94f8-5ffa1701a898","StatusNotification",{"connectorId":1,"errorCode":"NoError","status":"Charging","timestamp":"2024-11-27T00:35:51.815Z"}]
[MO] info (SmartChargingService.cpp:168): New limit for connector 1, scheduled at = 2024-11-27T00:35:51.897Z, nextChange = 2037-01-01T00:00:00.000Z, limit = {-1.0, -1.0, -1}
[MO] Recv: [3,"8712bef2-f7a9-8128-94f8-5ffa1701a898",{}]
[MO] Send: [2,"c5841c45-3f67-96db-d2cb-140cd44d3c24","MeterValues",{"connectorId":1,"transactionId":1732667752,"meterValue":[{"timestamp":"2024-11-27T00:36:51.895Z","sampledValue":[{"value":"0","context":"Sample.Periodic","measurand":"Energy.Active.Import.Register","unit":"Wh"},{"value":"0.00","context":"Sample.Periodic","measurand":"Power.Active.Import","unit":"W"},{"value":"0.00","context":"Sample.Periodic","measurand":"Current.Import","location":"Outlet","unit":"A"},{"value":"0.00","context":"Sample.Periodic","measurand":"Voltage","unit":"V"},{"value":"0.00","context":"Sample.Periodic","measurand":"SoC"}]}]}]
[MO] Recv: [3,"c5841c45-3f67-96db-d2cb-140cd44d3c24",{}]
[MO] Send: [2,"420e23d5-385c-3593-5639-334503f5d3a5","MeterValues",{"connectorId":1,"transactionId":1732667752,"meterValue":[{"timestamp":"2024-11-27T00:37:51.976Z","sampledValue":[{"value":"0","context":"Sample.Periodic","measurand":"Energy.Active.Import.Register","unit":"Wh"},{"value":"0.00","context":"Sample.Periodic","measurand":"Power.Active.Import","unit":"W"},{"value":"0.00","context":"Sample.Periodic","measurand":"Current.Import","location":"Outlet","unit":"A"},{"value":"0.00","context":"Sample.Periodic","measurand":"Voltage","unit":"V"},{"value":"0.00","context":"Sample.Periodic","measurand":"SoC"}]}]}]
```

**Reply by BJReplay** [reply]
Thanks.  I got there in the end, too, by giving up on secure, just going for IP addresses and insecure.

**Comment by purcell-lab** [comment]
I also add here that this OCPP integration also works well with iammeter OCPP charger simulator.

https://community.home-assistant.io/t/has-someone-controlled-the-ocpp-charger-to-consume-the-excess-solar-output/587271

![image](https://github.com/user-attachments/assets/3c4a3803-91e0-430d-99b3-5d0c82728730)

![image](https://github.com/user-attachments/assets/82b61307-e3e5-49a5-80fb-ec27470128ef)
